src/commons/sagas/NativeJsSaga.ts

```diff
@@ -0,0 +1,95 @@
+import { Context } from 'js-slang';
+import { ExceptionError } from 'js-slang/dist/errors/errors';
+import { SagaIterator } from 'redux-saga';
+import { call, put, select, takeEvery } from 'redux-saga/effects';
+import {
+  NATIVE_JS_RUN,
+  NativeJSEvalPayload,
+  NativeJSEvalResult
+} from 'src/features/nativeJS/NativeJSTypes';
+import { store } from 'src/pages/createStore';
+
+import { OverallState } from '../application/ApplicationTypes';
+import { evalNativeJSProgram } from '../nativeJS/NativeJSEval';
+import { actions } from '../utils/ActionsHelper';
+
+const dummyLocation = {
+  start: { line: 0, column: 0 },
+  end: { line: 0, column: 0 }
+};
+
+export function* nativeJsSaga(): SagaIterator {
+  yield takeEvery(NATIVE_JS_RUN, function* (action: ReturnType<typeof actions.nativeJSRun>) {
+    const { workspace, program }: NativeJSEvalPayload = action.payload;
+
+    // Notify workspace & clear REPL
+    yield put(actions.updateWorkspace(workspace, { isRunning: true }));
+    yield put(actions.clearReplOutput(workspace));
+    const context: Context = yield select(
+      (state: OverallState) => state.workspaces[workspace].context
+    );
+    context.errors = [];
+
+    async function evalWrapper() {
+      await new Promise(r => setTimeout(r, 0));
+      return call(evalNativeJSProgram, program);
+    }
+    const result: NativeJSEvalResult = yield (yield evalWrapper()) as Promise<NativeJSEvalResult>;
+
+    switch (result.status) {
+      case 'finished':
+        store.dispatch(actions.evalInterpreterSuccess(result.message, workspace));
+        break;
+      case 'error':
+        /**
+         * This function uses the non-standard feature `Error.prototype.stack`.
+         * Not working for SyntaxError.
+         *
+         * Some examples:
+         * `@http://localhost:8000/static/js/main.chunk.js line 31925 > eval:1:1
+         *     evalNativeJSProgram@http://localhost:8000/static/js/main.chunk.js:31925:16` --Firefox 91.4.1esr
+         * `TypeError: window.dsl is not a function at eval (eval at evalNativeJSProgram (NativeJSEval.ts:6:1), <anonymous>:1:8)` --Chrome 98.0.4758.9
+         * Seeking better alternative for this.
+         */
+        function getErrorLocation() {
```

Comment:
Should we put this under `/commons/nativeJS/NativeJSEval.ts` instead, since we will need to account for line number offsets for `eval` preludes? 

Then perhaps we can just return the final `ExceptionError` from `NativeJSEval.ts` also so that everything to do with evaluating and running the program string is under `NativeJSEval.ts` and let `NativeJsSaga.ts` handle the actions?

Reply:
Yes, this is a good idea. I wasn't thoughtful enough when writing these. Let's move it.